### PR TITLE
Adding tests for Bridge stuff that does not interact with Bridge API.

### PR DESCRIPTION
**Pull request by dsessler7**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?
   - [x] Have you added automated tests?

**Type of Changes:**

 
 - [ ] New feature
 - [ ] Bug fix
 - [ ] Documentation
 - [x] Testing enhancement
 - [ ] Other

**What is the current behavior (link to any open issues here)?**

**What is the new behavior (if this is a feature change)?**
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Other Information**:

### CODE REVIEW

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 1]
```diff
@@ -600,8 +513,8 @@ func (r *CrunchyBridgeClusterReconciler) GetSecretKeys(
 			return string(existing.Data["key"]), string(existing.Data["team"]), nil
 		}
 		err = fmt.Errorf("error handling secret: found key %t, found team %t",
```
Comment (jmckulk) [inline internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go:515]: It looks like this error message would be
> error handling secret: found key (true or false), found team (true or false)

I can't quite tell what is going on with this error message. Maybe say what we are expecting to find? :pick: 

something like ?
> expected to find a key and a team: found key true, found team false

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "secret-missing-api-key",
+				Namespace: ns,
+			},
+			Data: map[string][]byte{
+				"team": []byte(`jkl;`),
+			},
+		}
+		assert.NilError(t, tClient.Create(ctx, secret.DeepCopy()))
```
Comment (jmckulk) [inline internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go:64]: Do these need to be `DeepCopy` of the secrets? I see this being done a few times in this test.

I would expect a `DeepCopy` if we are updating the value of the secret and checking contents before and after

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+		apiKey, team, err := reconciler.GetSecretKeys(ctx, cluster)
+		assert.Check(t, apiKey == "")
+		assert.Check(t, team == "")
+		assert.ErrorContains(t, err, "error handling secret: found key false, found team true")
+
+		assert.NilError(t, tClient.Delete(ctx, secret.DeepCopy()))
```
Comment (jmckulk) [inline internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go:71]: I wouldn't think we'd want to be deleting a copy of an object :thinking:

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+		apiKey, team, err := reconciler.GetSecretKeys(ctx, cluster)
+		assert.Check(t, apiKey == "asdf")
+		assert.Check(t, team == "jkl;")
+		assert.NilError(t, err)
+	})
+
+}
+
+func TestDeleteControlled(t *testing.T) {
```
Comment (jmckulk) [inline internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go:115]: I see a test for `Owned` in our other `TestDeleteControlled` test. Should we have one here :question: 

https://github.com/CrunchyData/postgres-operator/blob/master/internal/controller/postgrescluster/controller_test.go#L79-L90

**Review by jmckulk — COMMENTED** [review]
Overall, the tests look good, but I do have some questions. I'm not too familiar with the bridge stuff, so definitely let me know if I'm missing something.

File: internal/bridge/crunchybridgecluster/helpers_test.go
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/helpers_test.go:1]: Yeahhh I thought about that, but 1) it felt a little weird to me to be including functions from the `postgrescluster` package in our `crunchybridgecluster` package and 2) I noticed that these functions were already duplicated in the `standalone_pgadmin` package, so I just did the same... Another option would be to pull them all out into their own package and include them as needed, but that feels like a large refactor that is outside the scope of this issue/story...

File: internal/bridge/crunchybridgecluster/watches.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crunchydata/postgres-operator/pkg/apis/postgres-operator.crunchydata.com/v1beta1"
+)
+
+// watchForRelatedSecret handles create/update/delete events for secrets,
+// passing the Secret ObjectKey to findCrunchyBridgeClustersForSecret
+func (r *CrunchyBridgeClusterReconciler) watchForRelatedSecret() handler.EventHandler {
+	handle := func(secret client.Object, q workqueue.RateLimitingInterface) {
+		ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/watches.go:34]: That's a good question. This was done by @benjaminjb  in the original iteration of the Bridge integration, so I'm not sure if there was specific reason for not using the global context...

File: internal/bridge/crunchybridgecluster/watches_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		assert.Equal(t, len(clusters), 0)
+	})
+
+	t.Run("OneCluster", func(t *testing.T) {
+		cluster1 := testCluster()
+		cluster1.Namespace = ns.Name
+		cluster1.Name = "first-cluster"
+		assert.NilError(t, tClient.Create(ctx, cluster1))
+
+		clusters := reconciler.findCrunchyBridgeClustersForSecret(ctx, secretObjectKey)
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/watches_test.go:59]: The secret name is defined by the `testCluster()` function call on line 54.

I can add another test case for making sure that a Bridge Cluster with a different secret name defined doesn't show up in the results.

File: internal/bridge/crunchybridgecluster/watches_test.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crunchydata/postgres-operator/internal/testing/require"
+)
+
+func TestFindCrunchyBridgeClustersForSecret(t *testing.T) {
+	ctx := context.Background()
+	_, tClient := setupKubernetes(t)
+	require.ParallelCapacity(t, 1)
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/watches_test.go:35]: TBH I added that simply because I saw it in our other `envtest` tests... I will try removing it.

File: internal/bridge/crunchybridgecluster/watches.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/watches.go:34]: (The contents of the `watches.go` file is "green" simply because I moved these functions from `crunchybridgecluster_controller.go` to their own file.)

File: internal/bridge/crunchybridgecluster/postgres_test.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	cluster.Namespace = setupNamespace(t, tClient).Name
+
+	spec := &v1beta1.CrunchyBridgeClusterRoleSpec{
+		Name:       "application",
+		SecretName: "application-role-secret",
+	}
+	role := &bridge.ClusterRoleApiResource{
+		Name:     "application",
+		Password: "password",
+		URI:      "postgres://application:password@p.33ddusyv3fgkrf2bz67bxlkpwa.db.postgresbridge.com:5432/postgres",
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/postgres_test.go:50]: It's an example host from the Bridge API docs, but yeah I can change it to something more generic/fake

File: internal/bridge/crunchybridgecluster/postgres_test.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+		spec2 := &v1beta1.CrunchyBridgeClusterRoleSpec{
+			Name:       "postgres",
+			SecretName: "role-secret",
+		}
+		cluster.Spec.Roles = append(cluster.Spec.Roles, spec1, spec2)
+
+		roleSpecSlice, secretMap, err := reconciler.reconcilePostgresRoleSecrets(ctx, apiKey, cluster)
+		assert.Check(t, roleSpecSlice == nil)
+		assert.Check(t, secretMap == nil)
+		assert.ErrorContains(t, err, "There are duplicate Role SecretNames in the spec. SecretNames must be unique.",
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/postgres_test.go:107]: So the message is saying that the CrunchyBridgeCluster spec has duplicate Role SecretNames (i.e. two or more Roles have the same SecretName)... I could maybe reword it to: "Two or more of the Roles in the CrunchyBridgeCluster spec have the same SecretName. Role SecretNames must be unique." How does that sound?

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go (hunk 4 above)
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go:115]: The test you are referring to tests the `setOwnerReference` function which we don't have/use in `crunchybridgecluster`.

File: internal/bridge/crunchybridgecluster/helpers_test.go
[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/helpers_test.go:1]: Agreed - a helper package sounds like a good idea but definitely out of the scope of this story. :+1:

[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/helpers_test.go:1]: Two other thoughts:
1. Could you add in the notes about these being duplicated functions - like we have in the standalone_pgadmin stuff?
2. Maybe separate out the `testCluster` function into its own file :thinking: so that if/when we make a new package the functions are nicely separated? Definitely not necessary but an idea

File: internal/bridge/crunchybridgecluster/postgres_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/postgres_test.go:107]: I think it would be fine to just add the type of spec to the original message.
```suggestion
		assert.ErrorContains(t, err, "There are duplicate Role SecretNames in the CrunchyBridgeCluster spec. SecretNames must be unique.",
```

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go (hunk 4 above)
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go:115]: Then should the first test be `NotControlled` instead of `NoOwnership`?

File: internal/bridge/crunchybridgecluster/watches.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/watches.go:34]: Yeah, I understand that this code is copied over. I just noticed it while I was looking through. I see other places where we use a Background context, so this is probably fine.

[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/watches.go:34]: Okay, I looked at this a bit and `watchForRelatedSecret` is called by `SetupWithManager` which does not define a context and doesn't accept one as an argument. None of our other `SetupWithManager` functions in our other controller packages have a context defined either. That doesn't necessarily mean we shouldn't do it here, but I presume that is why it was done this way.

File: internal/bridge/crunchybridgecluster/watches.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	return matching
+}
+
+//+kubebuilder:rbac:groups="postgres-operator.crunchydata.com",resources="crunchybridgeclusters",verbs={list}
+
+// Watch enqueues all existing CrunchyBridgeClusters for reconciles.
+func (r *CrunchyBridgeClusterReconciler) Watch() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(client.Object) []reconcile.Request {
+		ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/watches.go:94]: See responses to same question regarding `watchForRelatedSecret`.

File: internal/bridge/crunchybridgecluster/watches.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/watches.go:34]: Also, similarly in `postgrescluster`, `controllerRefHandlerFuncs` is called in `SetupWithManager` and it defines its own context rather than having one passed in as an argument.

[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/watches.go:34]: If we ever do want a context we can add it in :+1:

File: internal/bridge/crunchybridgecluster/watches_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/watches_test.go]: It looks like you removed `ParallelCapacity` rather than setting it to `0`. I see other new tests where it exists but is set to `0`. Do you have any idea what the difference between not running the function and setting it to `0` is?

Should we do one over the other?

[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/watches_test.go]: I think running it and setting it to `0` will signal that the test can be run in parallel. So i think we will want it to be set so that it can run in parallel with other tests.

https://github.com/CrunchyData/postgres-operator/blob/master/internal/testing/require/parallel.go#L29

**Review by jmckulk — APPROVED** [review]
